File: backend/app/routers/assets.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime

from app.db import get_db_session
from app.models.finance import Asset
from app.models.scenario import Scenario
from app.models.entity_value import EntityValue as EntityValueModel
from app.schemas.finance import AssetCreate, AssetUpdate, AssetRead
from app.schemas.entity_value import EntityValue
from app.routers.auth import get_current_user
from app.schemas import User


router = APIRouter(
    prefix="/assets",
    tags=["assets"],
    dependencies=[Depends(get_current_user)]
)


@router.post("/", response_model=AssetRead)
def create_asset(
    payload: AssetCreate,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Create a new asset."""
    asset = Asset(
        user_id=current_user.id,
        name=payload.name,
        asset_type=payload.asset_type,
        purchase_value=payload.purchase_value,
        purchase_date=payload.purchase_date,
        expected_annual_appreciation=payload.expected_annual_appreciation,
        is_primary_residence=payload.is_primary_residence,
        notes=payload.notes
    )
    db.add(asset)
    db.commit()
    db.refresh(asset)
    return asset


@router.get("/", response_model=List[AssetRead])
def get_assets(
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Get all assets for the current user."""
    assets = db.query(Asset).filter(
        Asset.user_id == current_user.id
    ).all()
    
    # Enrich assets with their latest values from the default scenario
    default_scenario = db.query(Scenario).filter(Scenario.name == "Actual").first()
    if not default_scenario:
        raise HTTPException(status_code=404, detail="Default scenario not found")
    
    for asset in assets:
        latest_value = asset.get_latest_value(db, default_scenario.id)
        if latest_value is not None:
            setattr(asset, "current_value", latest_value)
        else:
            setattr(asset, "current_value", 0.0)
    
    return assets


@router.get("/{asset_id}", response_model=AssetRead)
def get_asset(
    asset_id: int,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Get a specific asset."""
    asset = db.query(Asset).filter(
        Asset.id == asset_id,
        Asset.user_id == current_user.id
    ).first()
    
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    
    # Get the latest value from the default scenario
    default_scenario = db.query(Scenario).filter(Scenario.name == "Actual").first()
    if not default_scenario:
        raise HTTPException(status_code=404, detail="Default scenario not found")
    
    latest_value = asset.get_latest_value(db, default_scenario.id)
    if latest_value is not None:
        setattr(asset, "current_value", latest_value)
    else:
        setattr(asset, "current_value", 0.0)
    
    return asset


@router.put("/{asset_id}", response_model=AssetRead)
def update_asset(
    asset_id: int,
    payload: AssetUpdate,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Update an asset."""
    asset = db.query(Asset).filter(
        Asset.id == asset_id,
        Asset.user_id == current_user.id
    ).first()
    
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    
    # Update asset fields
    for field, value in payload.dict(exclude_unset=True).items():
        if hasattr(asset, field):
            setattr(asset, field, value)
    
    db.commit()
    db.refresh(asset)
    
    # Get the latest value from the default scenario
    default_scenario = db.query(Scenario).filter(Scenario.name == "Actual").first()
    if not default_scenario:
        raise HTTPException(status_code=404, detail="Default scenario not found")
    
    latest_value = asset.get_latest_value(db, default_scenario.id)
    if latest_value is not None:
        setattr(asset, "current_value", latest_value)
    else:
        setattr(asset, "current_value", 0.0)
    
    return asset


@router.post("/{asset_id}/value", response_model=EntityValue)
def set_asset_value(
    asset_id: int,
    value: float,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Set a new value for an asset."""
    asset = db.query(Asset).filter(
        Asset.id == asset_id,
        Asset.user_id == current_user.id
    ).first()
    
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    
    # Get the default scenario
    default_scenario = db.query(Scenario).filter(Scenario.name == "Actual").first()
    if not default_scenario:
        raise HTTPException(status_code=404, detail="Default scenario not found")
    
    # Create new entity value
    entity_value = asset.set_value(db, default_scenario.id, value)
    return entity_value


@router.delete("/{asset_id}")
def delete_asset(
    asset_id: int,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Delete an asset."""
    asset = db.query(Asset).filter(
        Asset.id == asset_id,
        Asset.user_id == current_user.id
    ).first()
    
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")
    
    # Delete associated entity values
    db.query(EntityValueModel).filter(
        EntityValueModel.entity_type == "ASSET",
        EntityValueModel.entity_id == asset_id
    ).delete()
    
    # Delete the asset
    db.delete(asset)
    db.commit()
    
    return {"message": "Asset deleted"} 

File: frontend/src/api/assets.ts
import { api } from './api';
import { Asset, EntityValue } from '@/types/finance';

export enum AssetTypeEnum {
  PRIMARY_RESIDENCE = "PRIMARY_RESIDENCE",
  SECONDARY_PROPERTY = "SECONDARY_PROPERTY",
  BUSINESS = "BUSINESS",
  VEHICLE = "VEHICLE",
  OTHER = "OTHER"
}

export interface Asset {
  id: number;
  name: string;
  asset_type: string;
  current_value: number;
  purchase_value?: number | null;
  purchase_date?: string | null;
  expected_annual_appreciation: number;
  is_primary_residence: boolean;
  notes: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface AssetCreate {
  name: string;
  asset_type: string;
  current_value: number;
  purchase_value?: number | null;
  purchase_date?: string | null;
  expected_annual_appreciation: number;
  is_primary_residence: boolean;
  notes?: string | null;
}

export interface AssetUpdate {
  name?: string;
  asset_type?: string;
  current_value?: number;
  purchase_value?: number | null;
  purchase_date?: string | null;
  expected_annual_appreciation?: number;
  is_primary_residence?: boolean;
  notes?: string | null;
}

// Helper function to ensure consistent asset structure
function normalizeAsset(asset: any): Asset {
  return {
    id: asset.id,
    name: asset.name,
    asset_type: asset.asset_type,
    current_value: typeof asset.current_value === 'number' ? asset.current_value : 0,
    purchase_value: asset.purchase_value || null,
    purchase_date: asset.purchase_date || null,
    expected_annual_appreciation: typeof asset.expected_annual_appreciation === 'number' ? asset.expected_annual_appreciation : 0,
    is_primary_residence: Boolean(asset.is_primary_residence),
    notes: asset.notes || null,
    created_at: asset.created_at,
    updated_at: asset.updated_at
  };
}

interface CreateAssetData {
  name: string;
  asset_type: string;
  initial_value?: number;
  expected_return_rate: number;
  notes?: string;
  family_member_id: number;
}

interface UpdateAssetData {
  name?: string;
  asset_type?: string;
  expected_return_rate?: number;
  notes?: string;
  family_member_id?: number;
}

export const assetsApi = {
  getAssets: async (): Promise<Asset[]> => {
    const response = await api.get<{ data: Asset[] }>('/assets');
    return response.data;
  },

  getAsset: async (id: number): Promise<Asset> => {
    const response = await api.get<{ data: Asset }>(`/assets/${id}`);
    return response.data;
  },

  createAsset: async (data: CreateAssetData): Promise<Asset> => {
    const response = await api.post<{ data: Asset }>('/assets', data);
    return response.data;
  },

  updateAsset: async (id: number, data: UpdateAssetData): Promise<Asset> => {
    const response = await api.put<{ data: Asset }>(`/assets/${id}`, data);
    return response.data;
  },

  deleteAsset: async (id: number): Promise<void> => {
    await api.delete(`/assets/${id}`);
  },

  setValue: async (id: number, value: number): Promise<EntityValue> => {
    const response = await api.post<{ data: EntityValue }>(`/assets/${id}/value`, { value });
    return response.data;
  }
}; 